### DISCUSSION TITLE
Predict model RE

### DISCUSSION THREAD

**Discussion by minhduc01168** [PaddlePaddle/PaddleOCR#14334] [Q&A]
Error:
File "/home/ubuntu/predict_paddle_kie/predict.py", line 296, in predict_and_save
    re_results, elapsed_time, ser_results = self.ser_re_engine(img, list_input_kie)
  File "/home/ubuntu/predict_paddle_kie/predict.py", line 263, in __call__
    re_input, entity_idx_dict_batch = make_input(ser_inputs, ser_results)
  File "/home/ubuntu/predict_paddle_kie/tools/infer/utility.py", line 763, in make_input
    assert len(entities) == len(ser_results)
AssertionError
I got the above error when doing predict. As far as I understand, ser_result and entities have an unequal number of elements? How can I fix this error?Can someone help me, thank you?

**Comment by GreatV** [comment]
Based on the error message and code analysis, I can help explain and fix the issue:

1. The error occurs because there's a mismatch between entities and ser_results lengths in the `make_input` function. The assertion `assert len(entities) == len(ser_results)` is failing.

2. This is likely happening because the data preprocessing for inference is not properly initialized. For RE prediction, you need both the SER results and the corresponding input data with entity information.

Here are a few ways to fix this:

1. Make sure your input data contains the required relations field:
```python
data = {
    "img_path": img_path,
    "relations": [],  # Add empty relations if not available 
    "label": label_info # If available
}
```

2. Modify the `SerRePredictor` calling code to properly initialize SER inputs:

```python
def predict_and_save(self, img):
    # First get SER results
    ser_results = self.ser_engine(img)
    
    # Make sure ser_results contains entity information
    if not ser_results or 'entities' not in ser_results[0]:
        return [], 0, []
        
    # Then process RE
    re_input, entity_idx_dict_batch = make_input(ser_inputs, ser_results)
    
    return re_results, elapsed_time, ser_results
```

3. If using the command line tool, make sure the config files for both SER and RE are properly set up:

```bash
python3 tools/infer_kie_token_ser_re.py \
  -c ./configs/kie/vi_layoutxlm/ser_RE_layoutxlm_xfund_zh.yml \
  -c_ser ./configs/kie/vi_layoutxlm/ser_layoutxlm_xfund_zh.yml \
  -o Architecture.Backbone.checkpoints=./pretrain_models/ser_layoutxlm_xfund_zh/best_accuracy \
  -o_ser Architecture.Backbone.checkpoints=./pretrain_models/ser_layoutxlm_xfund_zh/best_accuracy \
  Global.infer_img=./doc/imgs/test_vi_layout_1.jpg
```

The key points to ensure:

1. Your input image data is properly formatted
2. SER model produces valid entity information 
3. The config files contain all necessary fields for both SER and RE tasks
4. Checkpoint paths are correct for both models

If you're still having issues, try:
1. Print out `len(entities)` and `len(ser_results)` to see the actual mismatch
2. Debug the `make_input` function call to see what data is being passed
3. Verify your model checkpoints are compatible with the config

Let me know if you need any clarification or run into other issues!

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | claude-3-5-sonnet-latest*

**Reply by minhduc01168** [reply]
> Based on the error message and code analysis, I can help explain and fix the issue:
> 
> 1. The error occurs because there's a mismatch between entities and ser_results lengths in the `make_input` function. The assertion `assert len(entities) == len(ser_results)` is failing.
> 2. This is likely happening because the data preprocessing for inference is not properly initialized. For RE prediction, you need both the SER results and the corresponding input data with entity information.
> 
> Here are a few ways to fix this:
> 
> 1. Make sure your input data contains the required relations field:
> 
> ```python
> data = {
>     "img_path": img_path,
>     "relations": [],  # Add empty relations if not available 
>     "label": label_info # If available
> }
> ```
> 
> 2. Modify the `SerRePredictor` calling code to properly initialize SER inputs:
> 
> ```python
> def predict_and_save(self, img):
>     # First get SER results
>     ser_results = self.ser_engine(img)
>     
>     # Make sure ser_results contains entity information
>     if not ser_results or 'entities' not in ser_results[0]:
>         return [], 0, []
>         
>     # Then process RE
>     re_input, entity_idx_dict_batch = make_input(ser_inputs, ser_results)
>     
>     return re_results, elapsed_time, ser_results
> ```
> 
> 3. If using the command line tool, make sure the config files for both SER and RE are properly set up:
> 
> ```shell
> python3 tools/infer_kie_token_ser_re.py \
>   -c ./configs/kie/vi_layoutxlm/ser_RE_layoutxlm_xfund_zh.yml \
>   -c_ser ./configs/kie/vi_layoutxlm/ser_layoutxlm_xfund_zh.yml \
>   -o Architecture.Backbone.checkpoints=./pretrain_models/ser_layoutxlm_xfund_zh/best_accuracy \
>   -o_ser Architecture.Backbone.checkpoints=./pretrain_models/ser_layoutxlm_xfund_zh/best_accuracy \
>   Global.infer_img=./doc/imgs/test_vi_layout_1.jpg
> ```
> 
> The key points to ensure:
> 
> 1. Your input image data is properly formatted
> 2. SER model produces valid entity information
> 3. The config files contain all necessary fields for both SER and RE tasks
> 4. Checkpoint paths are correct for both models
> 
> If you're still having issues, try:
> 
> 1. Print out `len(entities)` and `len(ser_results)` to see the actual mismatch
> 2. Debug the `make_input` function call to see what data is being passed
> 3. Verify your model checkpoints are compatible with the config
> 
> Let me know if you need any clarification or run into other issues!
> 
> _Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | claude-3-5-sonnet-latest_

It seems your answer is not correct and has not solved the problem?

**Comment by GreatV** [comment]
It is not possible to obtain more useful information from your brief error message, so I can only speak in general terms. If you have a bug, please submit it to the **Issues** and provide more context.

**Comment by minhduc01168** [comment]
@GreatV I get an error at line: '''assert len(entities) == len(ser_results)''' when doing predict. The number of entities and ser_results are not equal when predicting. Specifically len(ser_results) > len(entities)?
def make_input(ser_inputs, ser_results):
    entities_labels = {"HEADER": 0, "QUESTION": 1, "ANSWER": 2}
    batch_size, max_seq_len = ser_inputs[0].shape[:2]
    entities = ser_inputs[8][0]
    ser_results = ser_results[0]
    assert len(entities) == len(ser_results)

    # entities
    start = []
    end = []
    label = []
    entity_idx_dict = {}
    for i, (res, entity) in enumerate(zip(ser_results, entities)):
        if res["pred"] == "O":
            continue
        entity_idx_dict[len(start)] = i
        start.append(entity["start"])
        end.append(entity["end"])
        label.append(entities_labels[res["pred"]])

    entities = np.full([max_seq_len + 1, 3], fill_value=-1, dtype=np.int64)
    entities[0, 0] = len(start)
    entities[1 : len(start) + 1, 0] = start
    entities[0, 1] = len(end)
    entities[1 : len(end) + 1, 1] = end
    entities[0, 2] = len(label)
    entities[1 : len(label) + 1, 2] = label

    # relations
    head = []
    tail = []
    for i in range(len(label)):
        for j in range(len(label)):
            if label[i] == 1 and label[j] == 2:
                head.append(i)
                tail.append(j)

    relations = np.full([len(head) + 1, 2], fill_value=-1, dtype=np.int64)
    relations[0, 0] = len(head)
    relations[1 : len(head) + 1, 0] = head
    relations[0, 1] = len(tail)
    relations[1 : len(tail) + 1, 1] = tail

    entities = np.expand_dims(entities, axis=0)
    entities = np.repeat(entities, batch_size, axis=0)
    relations = np.expand_dims(relations, axis=0)
    relations = np.repeat(relations, batch_size, axis=0)

    # remove ocr_info segment_offset_id and label in ser input
    if isinstance(ser_inputs[0], paddle.Tensor):
        entities = paddle.to_tensor(entities)
        relations = paddle.to_tensor(relations)
    ser_inputs = ser_inputs[:5] + [entities, relations]

    entity_idx_dict_batch = []
    for b in range(batch_size):
        entity_idx_dict_batch.append(entity_idx_dict)
    return ser_inputs, entity_idx_dict_batch

**Reply by GreatV** [reply]
Please provide the minimal reproducible demo

**Reply by minhduc01168** [reply]
@GreatV Can you tell me if len(entities) and len(ser_results) are required to be equal? In my opinion, after going through the SER model, some of the original entities have been separated, leading to len(ser_results) > len(entities)
